src/routing_core.rs

```diff
@@ -743,9 +743,313 @@ impl RoutingCore {
+                    for endpoint in connect_request.local_endpoints.iter() {
+                        if *endpoint == peer_endpoint {
+                            match value.0 {
+                                Some(_) => {
+                                    // If we've already matched a connection drop the new one.
+                                    let _ = self.action_sender.send(
+                                        ::action::Action::DropConnections(
+                                            vec![connection.clone()]));
+                                    return None
+                                },
+                                None => {
+                                    value.0 = Some(connection.clone());
+                                    let _ = self.action_sender.send(
+                                        ::action::Action::MatchConnection(
+                                            Some((key.clone(), value.0.clone())), None));
+                                    return Some(key.clone())
+                                }
+                            }
+                        }
+                    }
+                },
+                &ExpectedConnection::Response(ref connect_response, _) => {
+                    for endpoint in connect_response.local_endpoints.iter() {
+                        if *endpoint == peer_endpoint {
+                            match value.0 {
+                                Some(_) => {
+                                    // If we've already matched a connection drop the new one.
+                                    let _ = self.action_sender.send(
+                                        ::action::Action::DropConnections(
+                                            vec![connection.clone()]));
+                                    return None
+                                },
+                                None => {
+                                    value.0 = Some(connection.clone());
+                                    let _ = self.action_sender.send(
+                                        ::action::Action::MatchConnection(
+                                            Some((key.clone(), value.0.clone())), None));
+                                    return Some(key.clone())
+                                }
+                            }
+                        }
+                    }
+                },
+            }
+        }
+
+        None
+    }
+
+    /// Check whether the connection has been accepted.
+    pub fn match_unknown_connection(&mut self, connection: &::crust::Connection,
+            hello: &::direct_messages::Hello) {
+        for (key, value) in self.unknown_connections.iter_mut() {
+            if key == connection {
+                match value.0 {
+                    None => {
+                        value.0 = Some(hello.clone());
+                        let _ = self.action_sender.send(::action::Action::MatchConnection(
```

Comment:
Here we need to check whether it is a confirmation hello, consistent with the hello stored and continue to Match is so.